blogs/2021/09/06/bracket-pair-colorization.md

```diff
@@ -0,0 +1,311 @@
+
+How can we efficiently concatenize all these nodes of mixed heights into a single (2,3)-tree?
+This can easily be reduced to the problem of prepending or appending a smaller tree to a larger tree.
+If two trees have the same height, it is sufficient to create a list that contains both children.
+Otherwise, we insert the smaller tree of height $h_1$ into the larger tree of height $h_2$ and potentially break up nodes if they end up having more than 3 children (similar to how the insert operation of (2,3)-trees works). Because this has runtime $\mathcal{O}(h_2 - h_1)$, we take 3 adjacent nodes ($a$, $b$ and $c$) that we want to concatenize and concatenize either $a$ and $b$ or $b$ and $c$ first (potentially increasing the height of the tree), depending on which pair has the smaller height difference. This is repeated until all nodes are concatenized. As an additional optimization, we look for sequences of nodes that have the same height and create parent lists for them in linear time.
+
+To balance the lists α and γ of the previous example, we perform the concat operation on their children:
+
+![AST after balancing lists](./long-lists-tree-reuse3.dio.svg)
+
+Because list B has height 3 and bracket pair β height 0 in the unbalanced tree, we need to append β to B and are finished with list α. The remaining (2,3)-tree is B, thus it becomes the new root and replaces list α. Continuing with γ, its children δ and H have height 0, while G has height 1.
+We first concat δ and H and create a new parent node Y of height 1 (because δ and H have the same height). Then we concat Y and G and create a new parent list X (for the same reason). X then becomes the new child of the parent bracket pair, replacing the unbalanced list γ.
+
+We have to concatenize at most $\mathcal{O}(\mathrm{log}^2 N)$ many nodes with a maximum height of $\mathcal{O}(\mathrm{log} N)$ and additional $\mathcal{O}(E)$ many nodes of height 0. Because concatenizing two nodes of different height has time-complexity $\mathcal{O}(\mathrm{log} N)$ and all new nodes in a list are adjacent and have height 0, the time-complexity of the entire update operation is at most $\mathcal{O}(E + \mathrm{log}^3 N)$, given that finding a reusable node can be done fast enough.
+
+### How do we find reusable nodes efficiently?
+
+We have two data structures for this task: The *before edit position mapper* and the *node reader*.
+
+The [position mapper](https://github.com/microsoft/vscode/blob/f8e9f87b6554b527c61ba963d0c96c7687cbaae9/src/vs/editor/common/model/bracketPairColorizer/beforeEditPositionMapper.ts#L17) maps a position in the new document (after applying the edit) to the old document (before applying the edit), if possible. It also tells us the length between the current position and the next edit (or 0, if we are in an edit). This is done in $\mathcal{O}(1)$.
```

Comment:
I suggest linking to the code at the end. I'm also not 100% sure if these two paragraphs bring a lot of value, as they seem to describe the implementation in very high detail.